File: LAB_6_C#_C++/Expression.cs
using System;

namespace OOP_LAB_6
{
    class Expression
    {
        // Properties
        #region
        public double A { set; get; }
        public double B { set; get; }
        public double C { set; get; }
        public double D { set; get; }
        #endregion
        public Expression(double a, double b, double c, double d)
        {
            A = a;
            B = b;
            C = c;
            D = d;
        }

        public double Calculate()
        {
            if (24 + D - C < 0)
            {
                throw new InvalidOperationException();
            }
            if (Math.Sqrt(24 + D - C) + A/B == 0)
            {
                throw new DivideByZeroException();
            }

            return (1 + A - B / 2) / (Math.Sqrt(24 + D - C) + A / B);
        }

        public static Expression operator +(Expression exp, int num) => 
            new Expression(exp.A + num, exp.B + num, exp.C + num, exp.D + num);

        public static Expression operator +(Expression exp1, Expression exp2) => 
            new Expression(exp1.A + exp2.A, exp1.B + exp2.B, exp1.C + exp2.C, exp1.D + exp2.D);

        public static Expression operator -(Expression exp, int num) =>
            new Expression(exp.A - num, exp.B - num, exp.C - num, exp.D - num);

        public static Expression operator -(Expression exp1, Expression exp2) =>
            new Expression(exp1.A - exp2.A, exp1.B - exp2.B, exp1.C - exp2.C, exp1.D - exp2.D);

    }
}

File: LAB_6_C#_C++/Expression.cpp
#include "Expression.h"
#include <iostream>
#include <math.h>

using namespace std;

Expression::Expression(double a, double b, double c, double d)
{
	this->a = a;
	this->b = b;
	this->c = c;
	this->d = d;
}

double Expression::Calculate()
{
	if (24 + d - c < 0)
	{
		throw domain_error("Sqrt is less than 0");
	}
	if (sqrt(24 + d - c) + a / b == 0)
	{
		throw "Division by zero exception";
	}

	return (1 + a - b / 2) / (sqrt(24 + d - c) + a / b);
}

Expression* Expression::operator +(int num)
{
	return new Expression(this->a + num, this->b + num, this->c + num, this->d + num);
}

Expression* Expression::operator +(Expression exp2)
{
	return new Expression(this->a + exp2.a, this->b + exp2.b, this->c + exp2.c, this->d + exp2.d);
}

Expression* Expression::operator -(int num)
{
	return new Expression(this->a - num, this->b - num, this->c - num, this->d - num);
}

Expression* Expression::operator -(Expression exp2)
{
	return new Expression(this->a - exp2.a, this->b - exp2.b, this->c - exp2.c, this->d - exp2.d);
}